pyirf/irf/energy_dispersion.py

```diff
@@ -73,3 +75,86 @@ def energy_dispersion(
     energy_dispersion = _normalize_hist(energy_dispersion)
 
     return energy_dispersion
+
+
+def energy_dispersion_to_migration(
+    dispersion_matrix,
+    disp_true_energy_edges,
+    disp_migration_edges,
+    new_true_energy_edges,
+    new_reco_energy_edges,
+):
+    """
+    Construct a sparse energy migration matrix from a dense energy
+    dispersion matrix.
+    Depending on the new energy ranges, the sum over the first axis
+    can be smaller than 1.
+    The new true energy bins need to be a subset of the old range,
+    extrapolation is not supported.
+    New reconstruction bins outside of the old migration range are filled with
+    zeros.
+
+    Parameters
+    ----------
+    dispersion_matrix: numpy.ndarray
+        Energy dispersion_matrix
+    disp_true_energy_edges: astropy.units.Quantity[energy]
+        True energy edges matching the first dimension of the dispersion matrix
+    disp_migration_edges: numpy.ndarray
+        Migration edges matching the second dimension of the dispersion matrix
+    new_true_energy_edges: astropy.units.Quantity[energy]
+        True energy edges matching the first dimension of the output
+    new_reco_energy_edges: astropy.units.Quantity[energy]
+        Reco energy edges matching the second dimension of the output
+
+    Returns:
+    --------
+    migration_matrix: numpy.ndarray
+        Three-dimensional energy migration matrix. The third dimension
+        equals the fov offset dimension of the energy dispersion matrix.
+    """
+
+    migration_matrix = np.zeros((
+        len(new_true_energy_edges)-1,
+        len(new_reco_energy_edges)-1,
+        dispersion_matrix.shape[2],
+    ))
+
+    true_energy_interpolation = interpolate.interp1d(
+        (disp_true_energy_edges[1:] + disp_true_energy_edges[:-1])/2,
+        dispersion_matrix,
+        axis=0,
+    )
+    for idx, e_true in enumerate(
+            (new_true_energy_edges[1:] + new_true_energy_edges[:-1])/2
+    ):
+
+        # get migration for the new true energy bin
+        e_true_dispersion = true_energy_interpolation(e_true)
+
+        # for details of the interpolation see
+        # https://github.com/gammapy/gammapy/blob/f12ae6acece0ff5a6f269ca4f409f38c0714e23a/gammapy/irf/energy_dispersion.py#L242
+
+        # interpolate the cumsum of the dispersion values along the migration axis
+        cumsum = np.cumsum(e_true_dispersion, axis=0)
+        with np.errstate(invalid="ignore"):
+            cumsum = np.nan_to_num(cumsum / cumsum[-1])
+        f = interpolate.interp1d(
+            (disp_migration_edges[1:] + disp_migration_edges[:-1])/2,
```

Comment:
I think one should rather use there upper edges here, as the sum is defined up to the upper edge and no the center of the bin. Otherwise you shift the distribution by half an energy bin...

Reply:
Yes, I was confused about the desired dimensions with respect to the `np.insert`.
Edges should work better, I agree

Reply:
About the true energy loop:
Isn't this needed, because the target binning changes for each true energy?